File: app/src/test/java/cl/figonzal/evaluatool/evalua0/modulo1/SeriesE0M1Test.java
/*-----------------------------------------------------------------------------
                                                                              -
 This file is subject to the terms and conditions defined in                  -
 file 'LICENSE', which is part of this source code package                    -
                                                                              -
 Autor: Felipe González                                                       -
 Email: felipe.gonzalezalarcon94@gmail.com                                    -
                                                                              -
 Copyright (c) 2020                                                           -
                                                                              -
 Last modified 01-07-20 0:05                                                  -
 -----------------------------------------------------------------------------*/

package cl.figonzal.evaluatool.evalua0.modulo1;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

import java.util.Arrays;
import java.util.Collection;

import cl.figonzal.evaluatool.Utilidades;

import static org.junit.Assert.assertEquals;

@RunWith(Parameterized.class)
public class SeriesE0M1Test {

    private static final double DESVIACION = 9.68;
    private static final double MEDIA = 37.7;
    private final double percentil_esperado;
    private final double pd_total;
    private final double desviacion_esperada;
    private SeriesE0M1 m1;

    public SeriesE0M1Test(double pd_total, double percentil_esperado,
                          double desviacion_esperada) {
        this.percentil_esperado = percentil_esperado;
        this.pd_total = pd_total;
        this.desviacion_esperada = desviacion_esperada;
    }

    @Parameterized.Parameters()
    public static Collection<Object[]> data() {
        return Arrays.asList(new Object[][]{
                {48, 95, 1.06},
                {47, 90, 0.96},
                {46, 85, 0.86},
                {44, 80, 0.65},
                {42, 70, 0.44},
                {40, 60, 0.24},
                {38, 50, 0.03},
                {36, 45, -0.18},
                {34, 35, -0.38},
                {32, 30, -0.59},
                {30, 25, -0.80},
                {28, 20, -1.00},
                {26, 15, -1.21},
                {24, 12, -1.42},
                {22, 10, -1.62},
                {20, 9, -1.83},
                {18, 7, -2.04},
                {16, 5, -2.24},
                {14, 3, -2.45},
                {10, 1, -2.86},

                //LIMITES
                {9, 1, -2.96},
                {49, 95, 1.17}
        });
    }

    @Before
    public void setUp() {
        m1 = new SeriesE0M1();
    }

    @Test
    public void testCalcularPercentil() {
        assertEquals(percentil_esperado, m1.calcularPercentil(pd_total), 0.1);
    }

    @Test
    public void testCalcularDesviacion() {
        assertEquals(desviacion_esperada, Utilidades.calcularDesviacion(MEDIA, DESVIACION, pd_total, false), 0.001);
    }
}

File: app/src/test/java/cl/figonzal/evaluatool/ui/evaluas/evalua0/modulo1/SeriesE0M1Test.kt
/*

 This file is subject to the terms and conditions defined in
 file 'LICENSE', which is part of this source code package

 Autor: Felipe González
 Email: felipe.gonzalezalarcon94@gmail.com

 Copyright (c) 2023

 Last modified 18-06-23 18:48
 */
package cl.figonzal.evaluatool.ui.evaluas.evalua0.modulo1

import cl.figonzal.evaluatool.domain.baremo_tables.constants.Evalua0Constants.SERIES_E0M1
import cl.figonzal.evaluatool.domain.baremo_tables.tables.BaremoTable
import cl.figonzal.evaluatool.domain.baremo_tables.tables.Evalua0Baremo
import cl.figonzal.evaluatool.utils.EvaluaUtils.calculateDeviation
import cl.figonzal.evaluatool.utils.EvaluaUtils.calculatePercentile
import com.google.common.truth.Truth.assertThat
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.koin.core.context.startKoin
import org.koin.core.context.stopKoin
import org.koin.dsl.module
import org.koin.test.KoinTest
import org.koin.test.inject

class SeriesE0M1Test : KoinTest {

    private val baremoTable: BaremoTable by inject()
    private lateinit var perc: Array<DoubleArray>

    @Before
    fun setUp() {
        val startKoin = startKoin {

            modules(
                module {
                    single<BaremoTable> { Evalua0Baremo() }
                }
            )
        }

        perc = baremoTable.getBaremo(SERIES_E0M1)
    }

    @After
    fun cleanup() {
        stopKoin()
    }


    @Test
    fun testCalculatePercentile() {

        perc.forEach {

            val expPercentile = it[1].toInt()
            val calcPercentile = calculatePercentile(
                percentile = perc,
                pdTotal = it[0].toInt()
            )
            assertThat(expPercentile).isEqualTo(calcPercentile)
        }
    }

    @Test
    fun testCalculateDeviation() {

        perc.forEach {

            val expDeviation = it[2]
            val calcDeviation = calculateDeviation(
                MEAN = MEDIA,
                DEVIATION = DESVIACION,
                pdTotal = it[0].toInt()
            ).toDouble()
            assertThat(expDeviation).isEqualTo(calcDeviation)
        }
    }

    companion object {
        private const val DESVIACION = 9.68
        private const val MEDIA = 37.7
    }
}